RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc

```diff
@@ -168,185 +185,331 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
+            }
+            tmpCandidate.setPdgId(11 * track.charge());
+            float p = tracksterTotalRawPt * cosh(t.barycenter().eta());
+            tmpCandidate.setRawEnergy(p);
+            math::XYZTLorentzVector p4(
+                p * track.momentum().unit().x(), p * track.momentum().unit().y(), p * track.momentum().unit().z(), p);
+            tmpCandidate.setP4(p4);
+            resultCandidates->push_back(tmpCandidate);
+          }
+
+        } else {
+          // if 1-to-many find closest trackster in momentum
+          int closestTrackster = mergedIdx;
+          float minPtDiff = std::abs(t.raw_pt() - track.pt());
+          for (auto otherTracksterIdx : trackstersTRKEMwithSameSeed) {
+            auto thisPt = tracksterTotalRawPt + trackstersMergedHandle->at(otherTracksterIdx).raw_pt() - t.raw_pt();
+            closestTrackster = std::abs(thisPt - track.pt()) < minPtDiff ? otherTracksterIdx : closestTrackster;
+            usedTrackstersMerged[otherTracksterIdx] = true;
+          }
+          tracksterTotalRawPt += trackstersMergedHandle->at(closestTrackster).raw_pt() - t.raw_pt();
+
+          if (foundCompatibleTRK) {
+            TICLCandidate tmpCandidate;
+            tmpCandidate.addTrackster(edm::Ptr<ticl::Trackster>(trackstersMergedHandle, closestTrackster));
+            tmpCandidate.setCharge(track.charge());
+            tmpCandidate.setTrackPtr(edm::Ptr<reco::Track>(track_h, trackIdx));
+            tmpCandidate.setPdgId(211 * track.charge());
+            for (auto otherTracksterIdx : trackstersTRKwithSameSeed) {
+              tmpCandidate.addTrackster(edm::Ptr<ticl::Trackster>(trackstersMergedHandle, otherTracksterIdx));
+            }
+            float p = tracksterTotalRawPt * cosh(t.barycenter().eta());
+            float energy = std::sqrt(p * p + mpion2);
+            tmpCandidate.setRawEnergy(energy);
+            math::XYZTLorentzVector p4(p * track.momentum().unit().x(),
+                                       p * track.momentum().unit().y(),
+                                       p * track.momentum().unit().z(),
+                                       energy);
+            tmpCandidate.setP4(p4);
+            resultCandidates->push_back(tmpCandidate);
+
+          } else {
+            TICLCandidate tmpCandidate;
+            tmpCandidate.addTrackster(edm::Ptr<ticl::Trackster>(trackstersMergedHandle, closestTrackster));
+            tmpCandidate.setCharge(track.charge());
+            tmpCandidate.setTrackPtr(edm::Ptr<reco::Track>(track_h, trackIdx));
+            for (auto otherTracksterIdx : trackstersTRKwithSameSeed) {
+              tmpCandidate.addTrackster(edm::Ptr<ticl::Trackster>(trackstersMergedHandle, otherTracksterIdx));
+            }
+            tmpCandidate.setPdgId(11 * track.charge());
+            float p = tracksterTotalRawPt * cosh(t.barycenter().eta());
+            tmpCandidate.setRawEnergy(p);
```

Comment:
it seems like this line and the PDG ID are the only differences between this case and the `foundCompatibleTRK` case above. a lot of the duplication could be avoided. the same is likely also true for the other, similar block above.

Reply:
Since the PF interpretation is under active development, I would prefer that the factorization is postponed once this has become final and keep the code more readable (although longer).